src/theory/arith/rewriter/node_utils.h

```diff
@@ -0,0 +1,163 @@
+ */
+template <typename Iterable>
+std::optional<TNode> getZeroChild(const Iterable& parent)
+{
+  for (const auto& node : parent)
+  {
+    if (node.isConst() && node.template getConst<Rational>().isZero())
+    {
+      return node;
+    }
+  }
+  return {};
+}
+
+/** Create a Boolean constant node */
+inline Node mkConst(bool value)
+{
+  return NodeManager::currentNM()->mkConst(value);
+}
+/** Create an integer constant node */
+inline Node mkConst(const Integer& value)
+{
+  return NodeManager::currentNM()->mkConstInt(value);
+}
+/** Create an integer or rational constant node */
+inline Node mkConst(const Rational& value)
+{
+  if (value.isIntegral())
+  {
+    return NodeManager::currentNM()->mkConstInt(value);
+  }
+  return NodeManager::currentNM()->mkConstReal(value);
+}
+/** Create a real algebraic number node */
+inline Node mkConst(const RealAlgebraicNumber& value)
+{
+  return NodeManager::currentNM()->mkRealAlgebraicNumber(value);
+}
+
+/** Make a nonlinear multiplication from the given factors */
+inline Node mkNonlinearMult(const std::vector<Node>& factors)
+{
+  auto* nm = NodeManager::currentNM();
+  switch (factors.size())
+  {
+    case 0: return nm->mkConstInt(Rational(1));
+    case 1: return factors[0];
+    default: return nm->mkNode(Kind::NONLINEAR_MULT, factors);
+  }
+}
+
+/**
+ * Create the product of `multiplicity * monomial`. Assumes that the monomial is
+ * either a product of non-values (neither rational nor real algebraic numbers)
+ * or a rational constant.
+ * If the monomial is a constant, return the product of the two numbers. If the
+ * multiplicity is one, return the monomial. Otherwise return `(MULT
+ * multiplicity monomial)`.
+ */
+Node mkMultTerm(const Rational& multiplicity, TNode monomial);
+
+/**
+ * Create the product of `multiplicity * monomial`. Assumes that the monomial is
+ * either a product of non-values (neither rational nor real algebraic numbers)
+ * or a rational constant.
+ * If multiplicity is rational, defer to the appropriate overload. If the
+ * monomial is one, return the product of the two numbers. Otherwise return the
+ * nonlinear product of the two, i.e. `(NONLINEAR_MULT multiplicity *monomial)`.
+ */
+Node mkMultTerm(const RealAlgebraicNumber& multiplicity, TNode monomial);
+
+/**
+ * Create the product of `multiplicity * monomial`, where the monomial is given
+ * as the (implicitly multiplied, possibly unsorted) list of children. Assumes
+ * that monomial is either empty (implicitly one) or  a list of non-values
+ * (neither rational nor real algebraic numbers). If multiplicity is rational,
+ * sort the monomial, create a nonlinear mult term and defer to the appropriate
+ * overload. Otherwise return the nonlinear product of the two, i.e.
+ * `(NONLINEAR_MULT multiplicity *monomial)`. The monomial is taken as rvalue as
+ * it may be modified in the process.
+ *
+ */
+Node mkMultTerm(const RealAlgebraicNumber& multiplicity,
+                std::vector<Node>&& monomial);
```

Comment:
Why is this a double reference?

Reply:
The method possibly changes the monomial argument in a way that it does not make any sense to the caller. Thus, the caller should no longer use it afterwards. We therefore move it into the function (the `&&`), as noted in the comment.